VMEncryption/main/handle.py

```diff
@@ -666,6 +669,71 @@ def enable():
                           code=(CommonVariables.configuration_error),
                           message=msg)
 
+def are_required_devices_encrypted(volume_type, encryption_status, disk_util, bek_util, encryption_operation):
+    are_data_disk_encrypted = True if encryption_status['data'] == 'Encrypted' else False
+    is_os_disk_encrypted = True if encryption_status['os'] == 'Encrypted' else False
+    items_to_encrypt = []
+    if volume_type.lower() == CommonVariables.VolumeTypeData.lower():
+        if are_data_disk_encrypted:
+            logger.log('All data drives are encrypted.')
+            return True, items_to_encrypt
+        else:
+            logger.log('Not all data drives are encrypted.')
+            items_to_encrypt = find_all_devices_to_encrypt(None, disk_util, bek_util, volume_type, encryption_operation)
+            return False, items_to_encrypt
+    elif volume_type.lower() == CommonVariables.VolumeTypeOS.lower():
+        if is_os_disk_encrypted:
+            logger.log('OS drive is encrypted.')
+            return True, items_to_encrypt
+        else:
+            logger.log('OS Drive is not encrypted.')
+            items_to_encrypt = os_device_to_encrypt(disk_util)
+            return False, items_to_encrypt
+    elif volume_type.lower() == CommonVariables.VolumeTypeAll.lower():
+        if are_data_disk_encrypted and is_os_disk_encrypted:
+            logger.log('Both OS and Data drives are encrypted.')
+            return True, items_to_encrypt
+        else:
+            if not are_data_disk_encrypted:
+                items_to_encrypt = find_all_devices_to_encrypt(None, disk_util, bek_util, volume_type, encryption_operation)
+            if not is_os_disk_encrypted:
+                items_to_encrypt = items_to_encrypt + os_device_to_encrypt(disk_util)
+            return False, items_to_encrypt
+
+def handle_encryption(public_settings, encryption_status, disk_util, bek_util, encryption_operation):
+    extension_parameter = ExtensionParameter(hutil, logger, DistroPatcher, encryption_environment, get_protected_settings(), public_settings)
+    volume_type = public_settings.get(CommonVariables.VolumeTypeKey)
+
+    if extension_parameter.config_file_exists() and extension_parameter.config_changed():
+        logger.log("Config has changed, updating encryption settings")
+        are_devices_encrypted, items_to_encrypt = are_required_devices_encrypted(volume_type, encryption_status, disk_util, bek_util, encryption_operation)
+        if not are_devices_encrypted:
+            logger.log('Required devices not encrypted for volume type {0}. Calling update for already encrypted devices'.format(volume_type))
```

Comment:
Comment says - 'Calling update for already encrypted devices' but actually isn't it updating the encryption settings to include old devices plus new devices that are not already encrypted?

Reply:
Changed the log messages.